### ISSUE TITLE
[Question] Query-Builder and JSON:API Output

### ISSUE THREAD

**Issue by johannesschobel** [spatie/laravel-query-builder#100]
Dear Spatie Team,

thanks a lot for your awesome open-source packages - i really like the effort you put into your work! Keep it up - you're doing great..

This issue is not an issue - it is rather a general question on how to use this package in combination with "following the JSON:API standard" ( http://jsonapi.org/ )

From what I understand, this package (`Laravel-Query-Builder`) is used to automatically "map" request query parameters (e.g., `include / filter / sort / ...`) to a `QueryBuilder` instance that is used to select respective models from the database. In your `README` description you state, that these parameters (and the handling of the latter) closely follow the JSON:API specs - which i really (!) really like.. 

**BUT** - the description lacks information on returing data to the client. In your README you state:
> Once the relationships are loaded on the results collection you can include them in your response by using Eloquent API resources and conditional relationships.

However, the `Eloquent API Resources` **do not** allow for writing JSON:API compliant resources (at least not "out of the box). While it allows to "wrap" the content in a `data` field, it does not use the `data.attributes` and `data.type` fields, which should be added in order to comply with the standard. So there would be a huge overload to "re-build" the appropriate schema that is required to actually "use" JSON:API ... As far as i know, Laravels `Resources` do not have a concept like `Serializers` known from `Fractal`, which can change the "output format" automatically.

If i rely on `Fractal` (and your `Laravel-Fractal` wrappers) as described in your blog post here ( https://alexvanderbist.com/posts/2018/builder-apis-with-laravel-query-builder ) 
> Alternatively, you can use the tried and tested Fractal package by The League.
i would lose almost all features provided by this package.. But if i stick to `APIResources` from `Laravel` i would need to "re-structure" all `Resources` to follow the JSON:API standard..

So basically here is the question:
What do you suggest? How to use the `Query-Builder` **and** generate JSON:API compliant responses? Am I missing something or is it really a "either this or that" situation?!

All the best and thank you very much!

**Comment by ce-brex** [comment]
This is probably not the best avenue to discuss this, but having just completed a project using Query-Builder and the JSON:API format, I felt obligated to assist.  First off, this article helped me while I was implementing the JSON:API.

https://medium.com/@dinotedesco/using-laravel-5-5-resources-to-create-your-own-json-api-formatted-api-2c6af5e4d0e8

That should get you pointed in the right direction.  The API Resources allow you to transform the data being returned in the toArray() method.  So to wrap the resource, you would do something like 

```php
public function toArray($request) {
  return [
    'type' => 'resource-type',
    'id' => 'resource-id',
    'attributes' => [
      'name' => 'Some Name',
      'description' => 'Descriptive Text',
    ],
  ];
}
```

I did create some base classes to extend the default JsonResponse class to handle things like includes and such more easily, and then extended my classes with each API Resource.

**Comment by johannesschobel** [comment]
Hey @ce-brex ,

yeah, i know this article - and that is what really (!) buggles me.. Using `API Resources` requires you to **strictly** implement the JSON:API standard on your `toArray()` method. If you use - for example - the (famous) `Fractal` package, you do not need to "deeply nest" your own data into a `attributes` array, because a `Serializer` does this for you. So the example above would look like this (in Fractal):

```php
return [
   'id' => 'resource-id',
   'name' => 'Some Name',
   'description' => 'Descriptive Text',
];
```

Note that the `type` is **not** defined manually, but rather added implicit by configuration!

Then, a `JsonApiSerializer` is applied, that automatically transform it to a JSON:API valid output! However, `API Resources` from Laravel does not "know" such a concept - which makes it really (!) hard to apply some "standards".

The real benefit from `Fractal` in this case is, that the client communicating with your API may define the format he wants to have (e.g., via HTTP Headers). The API may then automatically switch the `Serializer` that is applied. That is **not** possible with `API Resources`, which - at least for me - is a real pain. From a RESTful perspective this really (!) tackles the overall principle of letting the client decide about the format and structure of the resources. For example, the client isn't even able to "change" the output format from JSON to XML (but who uses XML, right?! :laughing: ).

So - at least for me - using `API Resources` kind of feels wrong on so many different levels. But it could be that i am missing some really nice features that allow for that feature i really crave for.. Therefore, i wanted to know from the Spatie guys how they apply it in a JSON:API compliant way - because they advertise it as such!

All the best

**Comment by ce-brex** [comment]
Fractal is more of an all-in-one solution though.  Query-Builder is just for building the queries and supports the JSON:API spec as far as query parameters.  From there, it's easy to take the result and use it in an API Resource. I don't see where they advertise the result to be JSON:API compliant by itself.

**Comment by johannesschobel** [comment]
@ce-brex 
quote from the blog article i cited above in my first post of this issue:
> Finally, to make life even easier, the URL parameters in the package adhere to JSON API specification. (and so should you, your API's consumers will be forever grateful to you). JSON API specification describes some conventions and recommendations for JSON APIs. It will help you decide on what a request URL should look like and how a response might be formatted. This way you can focus on your business logic instead of what a filter query should look like.

**Comment by ce-brex** [comment]
Yes... the URL parameters, where it gets the information to build the query.

**Comment by johannesschobel** [comment]
yeah, i know.. But using the `Query-Builder` package here and then trying to use `Api Resources` feels kind of bad, because you need to **manually** build the JSON:API structure (e.g., adding the `attributes` "level" and manually maintaining the `type` and so on). Furthermore, you are "bound" to one specific structure that is "hard-coded" by you - you cannot easily switch the format based on clients requests..

If you try to use `Fractal` instead, there is a huge overlap in functionality..

Obviously, it would be better to stick with the "laravel style" (e.g., rely on `Api Resources`) but - at least in my opinion - they are not that powerful yet. So i wanted to know from the spatie guys how they would suggest to deal with this issue..

**Comment by freekmurze** [comment]
Building up responses is out of scope for this package. We only intend to build up the right query here. It feels like issue discussed should be submitted to Laravel itself.
